orchagent/fdborch.cpp

```diff
@@ -36,119 +38,145 @@ FdbOrch::FdbOrch(DBConnector *db, string tableName, PortsOrch *port) :
     Orch::addExecutor(fdbNotifier);
 }
 
-void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* entry, sai_object_id_t bridge_port_id)
+void FdbOrch::refreshFdbEntries()
 {
     SWSS_LOG_ENTER();
 
-    FdbUpdate update;
-    update.entry.mac = entry->mac_address;
-    update.entry.bv_id = entry->bv_id;
+    vector<string> keys;
+    m_fdbStateTable.getKeys(keys);
 
-    switch (type)
+    for (const auto &key: keys)
     {
-    case SAI_FDB_EVENT_LEARNED:
-        if (!m_portsOrch->getPortByBridgePortId(bridge_port_id, update.port))
+        std::vector<FieldValueTuple> fvs;
+        if (!m_fdbStateTable.get(key, fvs))
         {
-            SWSS_LOG_ERROR("Failed to get port by bridge port ID 0x%lx", bridge_port_id);
-            return;
+            SWSS_LOG_WARN("Empty field value tuple for %s", key.c_str());
+            continue;
         }
 
-        // we already have such entries
-        if (m_entries.find(update.entry) != m_entries.end())
+        sai_object_id_t bridge_port_id = SAI_NULL_OBJECT_ID;
+        for (auto &fv: fvs)
         {
-             SWSS_LOG_INFO("FdbOrch notification: mac %s is already in bv_id 0x%lx",
-                    update.entry.mac.to_string().c_str(), entry->bv_id);
-             break;
+            if(fvField(fv) == "SAI_FDB_ENTRY_ATTR_BRIDGE_PORT_ID")
+            {
+                sai_deserialize_object_id(fvValue(fv), bridge_port_id);
+            }
         }
 
-        update.add = true;
-
+        if (bridge_port_id == SAI_NULL_OBJECT_ID)
         {
-            auto ret = m_entries.insert(update.entry);
+            SWSS_LOG_WARN("Null bridge_port_id for %s", key.c_str());
+            continue;
+        }
 
-            SWSS_LOG_DEBUG("FdbOrch notification: mac %s was inserted into bv_id 0x%lx",
-                            update.entry.mac.to_string().c_str(), entry->bv_id);
+        sai_fdb_entry_t fdb_entry;
+        sai_deserialize_fdb_entry(key, fdb_entry);
+        this->update(SAI_FDB_EVENT_LEARNED, &fdb_entry, bridge_port_id);
+        SWSS_LOG_INFO("FDB from StateDB %s", key.c_str());
+    }
 
-            if (ret.second)
+    // TODO: do we need to sai_fdb_api->create_fdb_entry() during warm start?
+}
+
+void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* fdb_entry, sai_object_id_t bridge_port_id)
+{
```

Comment:
Did you make any function change in this function? Any issue with current code?

Reply:
It's just refactoring. No issue. I notice code duplication with FdbOrch::addFdbEntry(). So I make them very similar.


---
In reply to: [217211877](https://github.com/Azure/sonic-swss/pull/615#discussion_r217211877) [](ancestors = 217211877)